### ISSUE TITLE
patch: heap-buffer-overflow

### ISSUE THREAD

**Issue by 251** [landley/toybox#438]
Given file `A`:
```
$ hexdump -C A
00000000  2b 2b 2b 20 22 0f 00 00                           |+++ "...|
00000008
```

ASAN reports a heap-buffer-overflow with:
```
$ toybox-0.8.9/bin/toybox patch "" "A"
patch: =================================================================
==1301==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x602000000052 at pc 0x000000497dbc bp 0x7ffcfeb83dd0 sp 0x7ffcfeb83580
```

Looking at it with `gdb` it seems to scan past `filename`:
```
#0  __asan::ReportGenericError (pc=4816316, bp=bp@entry=140737488346896, sp=sp@entry=140737488344768, addr=105690555220050, is_write=false, access_size=3, exp=0, fatal=false) at /tmp/llvm-130/compiler-rt/lib/asan/asan_report.cpp:461
#1  0x0000000000497ddf in printf_common(void *, const char *, typedef __va_list_tag __va_list_tag *) (ctx=ctx@entry=0x7fffffffdf30, format=format@entry=0x52a33a "bad %s", aq=0x7fffffffdf38)
    at /tmp/llvm-130/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors_format.inc:553
#2  0x000000000049886f in __interceptor_vfprintf(__sanitizer::__sanitizer_FILE *, const char *, typedef __va_list_tag __va_list_tag *) (stream=0x7ffff7ea16a0 <_IO_2_1_stderr_>, format=<optimized out>, ap=0x7fffffffe060)
    at /tmp/llvm-130/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors.inc:1661
#3  0x00000000004f9a9c in verror_msg (msg=0x52a33a "bad %s", err=0, va=0x7fffffffe060) at lib/lib.c:14
#4  0x00000000004f9cab in error_exit (msg=0x52a33a "bad %s") at lib/lib.c:49
#5  0x0000000000511012 in unquote_file (filename=0x602000000050 "\276\276") at toys/posix/patch.c:289
#6  0x0000000000510619 in patch_main () at toys/posix/patch.c:387
```

(found by KLEE)

**Comment by landley** [comment]
What libc are you using? Because I _think_ libc it's what's scanning past the end there?

In patch.c unquote_file() receives basically a 3 byte string with a quote, a low ascii character, and a NUL terminator. (The second nul terminator is irrelevant-ish, patch doesn't handle embedded nuls so it acts as a truncated string.) It notices the starting quote and loops searching for the ending quote, hits the NUl first, and passes the input string (unmodified) to error_exit("bad %s", filename) because of the unbalanced quotes, which does a vfprintf(stderr, msg, ...) to basically pass it the "bad %s" and filename argument on to the library printf plumbing?

Both strings are null terminated, we allocated enough space for them, and . Unless we've corrupted the heap before now (which I assume your sanitizer would have caught) I don't see how the toybox code could control what libc's printf() does with those two arguments?

And if there _was_ something funky about how our filename() argument was allocated, your scrubber would have dinged upquote_file() traversing it before calling the error function? We called it because we found the terminating NUL without finding a matching quote, which means we read to the end of the string...

What did I miss? (Might still have some trailing dental drug residue, I was kind of out of it yesterday, but I'm not spotting it?)
